deeplinkdispatch/src/main/java/com/airbnb/deeplinkdispatch/BaseDeepLinkDelegate.kt

```diff
@@ -0,0 +1,442 @@
+     */
+    @JvmOverloads
+    fun dispatchFrom(
+        activity: Activity,
+        sourceIntent: Intent = activity.intent
+    ): DeepLinkResult {
+        val uri = sourceIntent.data
+        val result = uri?.let {
+            createResult(activity, sourceIntent, findEntry(uri.toString()))
+        } ?: createResult(activity, sourceIntent, null)
+        dispatchResult(
+            result = result,
+            activity = activity
+        )
+        notifyListener(
+            context = activity,
+            isError = !result.isSuccessful,
+            uri = uri,
+            uriTemplate = if (result.deepLinkMatchResult != null) result.deepLinkMatchResult.deeplinkEntry.uriTemplate else null,
+            errorMessage = result.error
+        )
+        return result
+    }
+
+    private fun notifyListener(
+        context: Context,
+        isError: Boolean,
+        uri: Uri?,
+        uriTemplate: String?,
+        errorMessage: String
+    ) {
+        val intent = Intent().apply {
+            action = DeepLinkHandler.ACTION
+            putExtra(DeepLinkHandler.EXTRA_URI, uri?.toString() ?: "")
+            putExtra(DeepLinkHandler.EXTRA_URI_TEMPLATE, uriTemplate ?: "")
+            putExtra(DeepLinkHandler.EXTRA_SUCCESSFUL, !isError)
+            if (isError) {
+                putExtra(DeepLinkHandler.EXTRA_ERROR_MESSAGE, errorMessage)
+            }
+        }
+        LocalBroadcastManager.getInstance(context).sendBroadcast(intent)
+    }
+
+    private fun dispatchResult(
+        result: DeepLinkResult,
+        activity: Activity
+    ) {
+        result.deepLinkMatchResult?.let { matchedDeepLinkResult ->
+            when (matchedDeepLinkResult.deeplinkEntry) {
+                is DeepLinkEntry.MethodDeeplinkEntry -> result.methodResult.taskStackBuilder?.startActivities()
+                    ?: activity.startActivity(result.methodResult.intent)
+                is DeepLinkEntry.ActivityDeeplinkEntry -> activity.startActivity(result.methodResult.intent)
+               is DeepLinkEntry.HandlerDeepLinkEntry -> if (result.isSuccessful) callDeeplinkHandler(result)
+            }
+        }
+    }
+
+    private fun callDeeplinkHandler(result: DeepLinkResult) {
+        val handlerClazz = result.deepLinkMatchResult?.deeplinkEntry?.clazz!!
+        val handlerMethod = handlerClazz.methods.singleOrNull { method ->
+            method.name == DEEP_LINK_HANDLER_METHOD_NAME && method.genericParameterTypes.let { parameterTypes ->
+                parameterTypes.singleOrNull()?.let { it != Object::class.java } ?: false
+            }
+        } ?: error("DeepLinkHandler class ${handlerClazz.name} has more than one methods " +
+                "\"$DEEP_LINK_HANDLER_METHOD_NAME\" methods with a single parameter.")
+        // Ok to call single here as we would have failed above if we would have more than one type.
+        val handlerParameterClazz = handlerMethod?.genericParameterTypes?.single() as Class<*>
```

Comment:
I don't think handlerMethod can be null here -- (as an aside, maybe we should configure kotlin errors as warnings for this repo?)